### ISSUE TITLE
Socket HTTP doesn't follow RFC specification

### ISSUE THREAD

**Issue by rjanja** [mafredri/openbttn#3]
Hi!

First: thank you so much for developing this firmware 👍! I've been looking for a way to reduce the latency between bttn push 🔴 and bttn feedback 🔊 for some time now. After flashing my bttn with `openbttn` I'm happy to say that my bttn presses now actually feel like they're doing something 🎉 😎. So thank you, thank you, thank you! 🍻

 I had a tough time trying to work with the bt.tn socket server. I ran into two main issues, which I'll detail below.

1. No examples for making requests to the socket server. This turned out to be trickier than I thought due to the way the shell passes newlines and carriage returns to `curl`. I think it'd be helpful to document this and possibly other ways of experimenting with the bt.tn.

```
export BTTNSOCK="http://10.1.5.51:8774/socket"
curl -w "\n" --data $'auth = PASSWORD\r\ndump_config' $BTTNSOCK
curl -w "\n" --data $'auth = PASSWORD\r\nblink_leds = ffffff;400;0;400' $BTTNSOCK
```

2. The socket server isn't outputting a CRLF between header and body and this causes most HTTP client libraries to fail with timeouts as they seem to continue waiting for a proper body to be sent from the server. The specification can be found at https://tools.ietf.org/html/rfc7230#section-3 and the culprit line would appear to be [src/wifi.c:285](https://github.com/mafredri/openbttn/blob/03fdf7dc1a45c0eb8cd3434f5ae661101cb70bc2/src/wifi.c#L285)

Something like this *should* work, but I may be wrong:
```diff
-"Content-Length: %d\n\n";
+"Content-Length: %d\r\n\r\n";
```
This was observed with Erlang's built-in `httpc` as well as `ibrowse` and `hackney`, where all connections timed out. Other clients should exhibit the same behavior, so I'm not sure why `curl` is working, unless it is just very lax with rules governing data formatting.

I was also able to channel requests through a proxy (Charles Proxy) to the socket server from all of the above clients with success, which made the problem harder to track down and also indicated the proxy was fixing-up the data from the server.

I've been able to work around this by implementing my own TCP client, but this might be a good candidate for fixing.

Let me know if you'd like a PR to document one or both of these things. I'm not in a great position to make modifications to the firmware, but could possibly set aside some time in a week or two for testing a change and re-flashing my bt.tn. I'm getting closer to an `openbttn` client library which will abstract some of this stuff from my applications, and I'll reach out to you for your approval before publishing. The audience may be pretty small, but we're a dedicated few, I think 😉 

Best,
Rob

**Comment by mafredri** [comment]
Hi @rjanja! Happy to hear you've found use for `openbttn`, it was born precisely out of the same frustration you were experiencing 😄.

> I had a tough time trying to work with the bt.tn socket server. I ran into two main issues, which I'll detail below.

Sorry about that, the socker server was originally created for communication with the WebUI and the led blinking got added as a quick demo hack 😅.

1. I think it's a good idea. We could definitely improve the documentation so that it's easier to test and experiment with the socket server. Did you find that you needed to use `\r\n` for the curl requests? I think I've mostly just used line feeds `\n` without the carriage return in my tests.

2. Oh, good catch. It should definitely be following spec!

> Something like this *should* work, but I may be wrong:

I think so too, although, we should probably use CRLF for all the headers, not just the body separator.

I'd definitely be happy to accept PRs if you wish to contribute, especially if you have ides for improving the documentaiton! That said, I could probably take a look at the header issue during this weekend and push an update for you to test 😄.

> I'm getting closer to an `openbttn` client library which will abstract some of this stuff from my applications, and I'll reach out to you for your approval before publishing.

That's awesome, I'm definitely looking forward to it 😁! Let me know if you have any ideas for improving the internals of the firmware to support this.

> The audience may be pretty small, but we're a dedicated few, I think 😉

Haha, you sure are right about that 😅.

---

Sidenote: There are a lot of improvements and cleanups to openbttn that I'd like to make, for example, I've been beta-testing a new firmware for the SPWF01SA Wi-Fi module which fixes a lot of bugs with the socket server that I've had to work around in the firmware. Not sure when it will be released though. I've also been entertaining rewriting the firmware in Rust, but we'll see what comes of that 😛.

**Comment by mafredri** [comment]
@rjanja I finally had time to test the header fix, and as far as I can tell it works fine 👍. I haven't verified Erlang-based http clients, I'm hoping you can verify that those are fixed now that #4 is merged.

Thanks for reporting this!

**Comment by rjanja** [comment]
Hi @mafredri!

Thanks for the quick response. I apologize it's taken me some time to get back to this. But I can now report that your header fix is working as expected with Erlang `ibrowse` and `hackney` clients, and I've been able to scrap the low-level socket connection workaround that I was previously using to communicate with my bttn.

I've taken a stab at a minor documentation update and will open a PR separately for your review. Feel free to wrangle in what you'd like, if anything, however you see fit.

The two things I'm most interested in seeing are now noted in the README:
- explanation of how LEDs can be addressed and controlled (I'm not a hardware guy)
- more info about the Wi-Fi module update (I haven't applied it, not sure if I should attempt to)

Thanks again for addressing this so quickly!
